File: code/178/solution.go
package main /*{{{*/

import (
	"fmt"
	"math"
)

const (
	e4  = int64(10000)
	e5  = int64(100000)
	e6  = int64(1000000)
	e7  = int64(10000000)
	e8  = e4 * e4
	mod = int64(1e9 + 7)
)

func sqrt(a int64) int64 {
	x := math.Sqrt(float64(a))
	return int64(x)
}

/*}}}*/

var cnt = [50][11][1050]int64{}

func main() {
	for i := 1; i <= 9; i++ {
		cnt[1][i][1<<i] = 1
	}
	for i := 1; i < 40; i++ {
		for j := 0; j <= 9; j++ {
			for k := 0; k < 1024; k++ {
				if j > 0 {
					cnt[i+1][j-1][k|1<<(j-1)] += cnt[i][j][k]
				}
				if j < 9 {
					cnt[i+1][j+1][k|1<<(j+1)] += cnt[i][j][k]
				}
			}
		}
	}
	ans := int64(0)
	for i := 10; i <= 40; i++ {
		for j := 0; j < 10; j++ {
			ans += cnt[i][j][1023]
		}
	}
	fmt.Println(ans)
}

File: code/178/solution.cpp
#include "../template/bigint.cpp"
#include "../template/china_reminder.cpp"
#include "../template/euler.cpp"
#include "../template/pell.cpp"
#include "../template/prime.cpp"
#include "../template/prime_cnt.cpp"
#include <iostream>/*{{{*/
#include <map>
#include <math.h>
#include <set>
#include <vector>
const long long e2 = 100;
const long long e3 = e2 * 10;
const long long e4 = e3 * 10;
const long long e5 = e4 * 10;
const long long e6 = e4 * e2;
const long long e7 = e6 * 10;
const long long e8 = e7 * 10;
const long long e9 = e8 * 10;
long long _mod     = e9 + 7;
using namespace std; /*}}}*/

long long cnt[50][11][1050]; // len, tail, acc
int main() {
    for (int i = 1; i <= 9; i++)
        cnt[1][i][1 << i] = 1;
    for (int i = 1; i < 40; i++) {
        for (int j = 0; j <= 9; j++) {
            for (int k = 0; k < 1024; k++) {
                if (j > 0)
                    cnt[i + 1][j - 1][k | 1 << (j - 1)] += cnt[i][j][k];
                if (j < 9)
                    cnt[i + 1][j + 1][k | 1 << (j + 1)] += cnt[i][j][k];
            }
        }
    }
    long long ans = 0;
    for (int i = 10; i <= 40; i++) {
        for (int j = 0; j <= 9; j++) {
            ans += cnt[i][j][1023];
        }
    }
    cout << ans << endl;
    return 0;
}
